File: examples/translation/translation.py
#!/usr/bin/env python3
"""
Even G2 Translation - Real-time Speech Translation

Enables translation mode on G2 glasses and displays translation results.

Usage:
    python translation.py CS EN     # Czech to English
    python translation.py HK EN     # Cantonese to English
    python translation.py --list    # Show available languages

Requirements:
    pip install bleak
"""

import asyncio
import sys
import time
from bleak import BleakClient, BleakScanner
from enum import Enum
from typing import Optional

# BLE UUIDs for Even G2
UUID_BASE = "00002760-08c2-11e1-9073-0e8ac72e{:04x}"
CHAR_WRITE = UUID_BASE.format(0x5401)
CHAR_NOTIFY = UUID_BASE.format(0x5402)


# Common language codes
LANGUAGES = {
    'EN': 'English',
    'CS': 'Czech',
    'HK': 'Cantonese (Hong Kong)',
    'ZH': 'Mandarin Chinese',
    'JA': 'Japanese',
    'KO': 'Korean',
    'ES': 'Spanish',
    'FR': 'French',
    'DE': 'German',
    'IT': 'Italian',
    'PT': 'Portuguese',
    'RU': 'Russian',
    'AR': 'Arabic',
}


def crc16_ccitt(data: bytes) -> int:
    """CRC-16/CCITT for packet framing."""
    crc = 0xFFFF
    for byte in data:
        crc ^= byte << 8
        for _ in range(8):
            crc = ((crc << 1) ^ 0x1021) if crc & 0x8000 else (crc << 1)
            crc &= 0xFFFF
    return crc


def encode_varint(value: int) -> bytes:
    """Encode integer as protobuf varint."""
    result = []
    while value > 0x7F:
        result.append((value & 0x7F) | 0x80)
        value >>= 7
    result.append(value & 0x7F)
    return bytes(result)


def build_packet(seq: int, svc_hi: int, svc_lo: int, payload: bytes) -> bytes:
    """Build a complete G2 packet with header and CRC."""
    header = bytes([0xAA, 0x21, seq, len(payload) + 2, 0x01, 0x01, svc_hi, svc_lo])
    full = header + payload
    crc = crc16_ccitt(payload)
    return full + bytes([crc & 0xFF, (crc >> 8) & 0xFF])


def build_auth_packets() -> list:
    """Build the 7-packet authentication sequence."""
    timestamp = int(time.time())
    ts_varint = encode_varint(timestamp)
    txid = bytes([0xE8, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0x01])
    
    def add_crc(packet: bytes) -> bytes:
        crc = crc16_ccitt(packet[8:])
        return packet + bytes([crc & 0xFF, (crc >> 8) & 0xFF])
    
    packets = []
    
    packets.append(add_crc(bytes([
        0xAA, 0x21, 0x01, 0x0C, 0x01, 0x01, 0x80, 0x00,
        0x08, 0x04, 0x10, 0x0C, 0x1A, 0x04, 0x08, 0x01, 0x10, 0x04
    ])))
    packets.append(add_crc(bytes([
        0xAA, 0x21, 0x02, 0x0A, 0x01, 0x01, 0x80, 0x20,
        0x08, 0x05, 0x10, 0x0E, 0x22, 0x02, 0x08, 0x02
    ])))
    
    payload3 = bytes([0x08, 0x80, 0x01, 0x10, 0x0F, 0x82, 0x08, 0x11, 0x08]) + ts_varint + bytes([0x10]) + txid
    packets.append(add_crc(bytes([0xAA, 0x21, 0x03, len(payload3) + 2, 0x01, 0x01, 0x80, 0x20]) + payload3))
    
    packets.append(add_crc(bytes([
        0xAA, 0x21, 0x04, 0x0C, 0x01, 0x01, 0x80, 0x00,
        0x08, 0x04, 0x10, 0x10, 0x1A, 0x04, 0x08, 0x01, 0x10, 0x04
    ])))
    packets.append(add_crc(bytes([
        0xAA, 0x21, 0x05, 0x0C, 0x01, 0x01, 0x80, 0x00,
        0x08, 0x04, 0x10, 0x11, 0x1A, 0x04, 0x08, 0x01, 0x10, 0x04
    ])))
    packets.append(add_crc(bytes([
        0xAA, 0x21, 0x06, 0x0A, 0x01, 0x01, 0x80, 0x20,
        0x08, 0x05, 0x10, 0x12, 0x22, 0x02, 0x08, 0x01
    ])))
    
    payload7 = bytes([0x08, 0x80, 0x01, 0x10, 0x13, 0x82, 0x08, 0x11, 0x08]) + ts_varint + bytes([0x10]) + txid
    packets.append(add_crc(bytes([0xAA, 0x21, 0x07, len(payload7) + 2, 0x01, 0x01, 0x80, 0x20]) + payload7))
    
    return packets


def build_translation_enable(seq: int, msg_id: int, source: str, target: str) -> bytes:
    """
    Build packet to enable translation with specified language pair.
    
    Args:
        seq: Packet sequence number
        msg_id: Message ID
        source: Source language code (e.g., "CS")
        target: Target language code (e.g., "EN")
    
    Returns:
        Complete packet bytes
    """
    lang_pair = f"{source}>{target}".encode('utf-8')
    
    # Mode data: 08 01 12 len lang 18 01
    mode_data = bytes([0x08, 0x01, 0x12, len(lang_pair)]) + lang_pair + bytes([0x18, 0x01])
    
    # Payload: 08 01 10 msg_id 1A len mode_data
    payload = bytes([0x08, 0x01, 0x10, msg_id, 0x1A, len(mode_data)]) + mode_data
    
    return build_packet(seq, 0x05, 0x20, payload)


def build_translation_disable(seq: int, msg_id: int) -> bytes:
    """
    Build packet to disable translation.
    
    Args:
        seq: Packet sequence number
        msg_id: Message ID
    
    Returns:
        Complete packet bytes
    """
    # Mode data: 08 02
    mode_data = bytes([0x08, 0x02])
    
    # Payload: 08 01 10 msg_id 1A len mode_data
    payload = bytes([0x08, 0x01, 0x10, msg_id, 0x1A, len(mode_data)]) + mode_data
    
    return build_packet(seq, 0x05, 0x20, payload)


def parse_translation_result(data: bytes) -> Optional[dict]:
    """
    Parse translation result from notification data.
    
    Returns dict with keys: original, translation, is_final
    """
    # Skip header (AA 21 seq len 01 01 05 20)
    if len(data) < 10:
        return None
    
    # Find service 05 20
    try:
        idx = data.index(bytes([0x05, 0x20]))
        payload = data[idx + 2:-2]  # Skip service and CRC
    except ValueError:
        return None
    
    if len(payload) < 4 or payload[0] != 0x08 or payload[1] != 0x02:
        return None
    
    result = {}
    idx = 2
    
    while idx < len(payload):
        tag = payload[idx]
        
        if tag == 0x10:  # Message ID
            idx += 1
            if idx < len(payload):
                result['msg_id'] = payload[idx]
                idx += 1
        
        elif tag == 0x22:  # Translation content
            idx += 1
            if idx >= len(payload):
                break
            content_len = payload[idx]
            idx += 1
            content = payload[idx:idx + content_len]
            idx += content_len
            
            # Parse content: 0A len original 12 len translation
            cidx = 0
            if cidx < len(content) and content[cidx] == 0x0A:
                cidx += 1
                if cidx >= len(content):
                    continue
                orig_len = content[cidx]
                cidx += 1
                result['original'] = content[cidx:cidx + orig_len].decode('utf-8', errors='replace')
                cidx += orig_len
                
                if cidx < len(content) and content[cidx] == 0x12:
                    cidx += 1
                    if cidx >= len(content):
                        continue
                    trans_len = content[cidx]
                    cidx += 1
                    result['translation'] = content[cidx:cidx + trans_len].decode('utf-8', errors='replace')
        
        elif tag == 0x18:
            idx += 2
        
        elif tag == 0x20:  # Final flag
            idx += 1
            if idx < len(payload):
                result['is_final'] = payload[idx] == 0x01
                idx += 1
        
        elif tag == 0x2A:  # Speaker info
            idx += 1
            if idx >= len(payload):
                break
            spk_len = payload[idx]
            idx += 1
            idx += spk_len
        
        else:
            idx += 1
    
    return result if 'original' in result else None


class TranslationHandler:
    """Handles translation session with G2 glasses."""
    
    def __init__(self, source: str, target: str):
        self.source = source.upper()
        self.target = target.upper()
        self.seq = 0x10
        self.msg_id = 0x50
        self.translation_count = 0
    
    def handle_notification(self, sender, data: bytes):
        """Handle BLE notification data."""
        result = parse_translation_result(data)
        if result:
            self.translation_count += 1
            final = "[FINAL]" if result.get('is_final') else "[...]"
            print(f"\n{final} Original: {result.get('original', '')}")
            print(f"      Translation: {result.get('translation', '')}")


async def main():
    args = sys.argv[1:]
    
    if '--list' in args or '-l' in args:
        print("Available language codes:")
        print("-" * 40)
        for code, name in LANGUAGES.items():
            print(f"  {code}: {name}")
        print("\nUsage: python translation.py SOURCE TARGET")
        print("Example: python translation.py CS EN")
        return
    
    if len(args) < 2:
        print("Even G2 Translation")
        print("=" * 40)
        print("\nUsage:")
        print("  python translation.py SOURCE TARGET")
        print("  python translation.py --list")
        print("\nExamples:")
        print("  python translation.py CS EN    # Czech to English")
        print("  python translation.py HK EN    # Cantonese to English")
        return
    
    source = args[0].upper()
    target = args[1].upper()
    
    print("Even G2 Translation")
    print("=" * 40)
    print(f"\nLanguage: {LANGUAGES.get(source, source)} → {LANGUAGES.get(target, target)}")
    
    print("\nScanning for G2 glasses...")
    devices = await BleakScanner.discover(timeout=10.0)
    
    # Find left eye (primary)
    g2_device = next(
        (d for d in devices if d.name and "G2" in d.name and "_L_" in d.name),
        None
    )
    
    if not g2_device:
        g2_device = next(
            (d for d in devices if d.name and "G2" in d.name),
            None
        )
    
    if not g2_device:
        print("ERROR: No G2 glasses found!")
        return
    
    print(f"  Found: {g2_device.name}")
    
    handler = TranslationHandler(source, target)
    
    async with BleakClient(g2_device) as client:
        print("\nConnected!")
        
        # Enable notifications
        await client.start_notify(CHAR_NOTIFY, handler.handle_notification)
        
        # Authenticate
        print("\nAuthenticating...")
        for pkt in build_auth_packets():
            await client.write_gatt_char(CHAR_WRITE, pkt, response=False)
            await asyncio.sleep(0.1)
        await asyncio.sleep(0.5)
        print("  Authenticated!")
        
        # Enable translation
        print(f"\nEnabling {source}>{target} translation...")
        enable_pkt = build_translation_enable(handler.seq, handler.msg_id, source, target)
        await client.write_gatt_char(CHAR_WRITE, enable_pkt, response=False)
        handler.seq += 1
        handler.msg_id += 1
        
        print("\nTranslation active! Speak to see results.")
        print("Press Ctrl+C to stop.\n")
        
        try:
            while True:
                await asyncio.sleep(1)
        except asyncio.CancelledError:
            pass
        finally:
            # Disable translation
            print("\n\nDisabling translation...")
            disable_pkt = build_translation_disable(handler.seq, handler.msg_id)
            await client.write_gatt_char(CHAR_WRITE, disable_pkt, response=False)
            await asyncio.sleep(0.5)
            print(f"Done! Received {handler.translation_count} translation(s).")


if __name__ == "__main__":
    try:
        asyncio.run(main())
    except KeyboardInterrupt:
        print("\nInterrupted")

File: examples/rust/src/bin/translation.rs
//! Even G2 Translation - Real-time Speech Translation
//!
//! Enables translation mode on G2 glasses and displays translation results.
//!
//! # Usage
//!
//! ```bash
//! cargo run --bin translation -- CS EN     # Czech to English
//! cargo run --bin translation -- HK EN     # Cantonese to English
//! cargo run --bin translation -- --list    # Show available languages
//! ```
//!
//! # Requirements
//!
//! - Bluetooth adapter with BLE support
//! - Even G2 glasses

use btleplug::api::{Central, Manager as _, Peripheral as _, ScanFilter, WriteType};
use btleplug::platform::{Adapter, Manager, Peripheral};
use futures::stream::StreamExt;
use std::collections::HashMap;
use std::env;
use std::sync::{Arc, Mutex};
use std::time::Duration;
use tokio::time::sleep;
use uuid::Uuid;

// BLE UUIDs
const UUID_BASE: &str = "00002760-08c2-11e1-9073-0e8ac72e";
fn char_write() -> Uuid {
    Uuid::parse_str(&format!("{}{:04x}", UUID_BASE, 0x5401)).unwrap()
}
fn char_notify() -> Uuid {
    Uuid::parse_str(&format!("{}{:04x}", UUID_BASE, 0x5402)).unwrap()
}

/// CRC-16/CCITT for packet framing
fn crc16_ccitt(data: &[u8]) -> u16 {
    let mut crc: u16 = 0xFFFF;
    for &byte in data {
        crc ^= (byte as u16) << 8;
        for _ in 0..8 {
            if crc & 0x8000 != 0 {
                crc = ((crc << 1) ^ 0x1021) & 0xFFFF;
            } else {
                crc = (crc << 1) & 0xFFFF;
            }
        }
    }
    crc
}

/// Encode integer as protobuf varint
fn encode_varint(mut value: u64) -> Vec<u8> {
    let mut result = Vec::new();
    while value > 0x7F {
        result.push(((value & 0x7F) | 0x80) as u8);
        value >>= 7;
    }
    result.push((value & 0x7F) as u8);
    result
}

/// Build a complete packet with header and CRC
fn build_packet(seq: u8, svc_hi: u8, svc_lo: u8, payload: &[u8]) -> Vec<u8> {
    let mut packet = vec![
        0xAA,
        0x21,
        seq,
        (payload.len() + 2) as u8,
        0x01,
        0x01,
        svc_hi,
        svc_lo,
    ];
    packet.extend_from_slice(payload);
    let crc = crc16_ccitt(payload);
    packet.push((crc & 0xFF) as u8);
    packet.push(((crc >> 8) & 0xFF) as u8);
    packet
}

/// Add CRC to packet
fn add_crc(mut packet: Vec<u8>) -> Vec<u8> {
    let crc = crc16_ccitt(&packet[8..]);
    packet.push((crc & 0xFF) as u8);
    packet.push(((crc >> 8) & 0xFF) as u8);
    packet
}

/// Build authentication packets
fn build_auth_packets() -> Vec<Vec<u8>> {
    let timestamp = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .unwrap()
        .as_secs();
    let ts_varint = encode_varint(timestamp);
    let txid = vec![0xE8, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0x01];

    let mut packets = Vec::new();

    packets.push(add_crc(vec![
        0xAA, 0x21, 0x01, 0x0C, 0x01, 0x01, 0x80, 0x00, 0x08, 0x04, 0x10, 0x0C, 0x1A, 0x04, 0x08,
        0x01, 0x10, 0x04,
    ]));

    packets.push(add_crc(vec![
        0xAA, 0x21, 0x02, 0x0A, 0x01, 0x01, 0x80, 0x20, 0x08, 0x05, 0x10, 0x0E, 0x22, 0x02, 0x08,
        0x02,
    ]));

    let mut payload = vec![0x08, 0x80, 0x01, 0x10, 0x0F, 0x82, 0x08, 0x11, 0x08];
    payload.extend_from_slice(&ts_varint);
    payload.push(0x10);
    payload.extend_from_slice(&txid);
    let mut pkt = vec![0xAA, 0x21, 0x03, (payload.len() + 2) as u8, 0x01, 0x01, 0x80, 0x20];
    pkt.extend_from_slice(&payload);
    packets.push(add_crc(pkt));

    packets.push(add_crc(vec![
        0xAA, 0x21, 0x04, 0x0C, 0x01, 0x01, 0x80, 0x00, 0x08, 0x04, 0x10, 0x10, 0x1A, 0x04, 0x08,
        0x01, 0x10, 0x04,
    ]));
    packets.push(add_crc(vec![
        0xAA, 0x21, 0x05, 0x0C, 0x01, 0x01, 0x80, 0x00, 0x08, 0x04, 0x10, 0x11, 0x1A, 0x04, 0x08,
        0x01, 0x10, 0x04,
    ]));

    packets.push(add_crc(vec![
        0xAA, 0x21, 0x06, 0x0A, 0x01, 0x01, 0x80, 0x20, 0x08, 0x05, 0x10, 0x12, 0x22, 0x02, 0x08,
        0x01,
    ]));

    let mut payload = vec![0x08, 0x80, 0x01, 0x10, 0x13, 0x82, 0x08, 0x11, 0x08];
    payload.extend_from_slice(&ts_varint);
    payload.push(0x10);
    payload.extend_from_slice(&txid);
    let mut pkt = vec![0xAA, 0x21, 0x07, (payload.len() + 2) as u8, 0x01, 0x01, 0x80, 0x20];
    pkt.extend_from_slice(&payload);
    packets.push(add_crc(pkt));

    packets
}

/// Build translation enable packet
pub fn build_translation_enable(seq: u8, msg_id: u8, source: &str, target: &str) -> Vec<u8> {
    let lang_pair = format!("{}>{}", source, target);
    let lang_bytes = lang_pair.as_bytes();

    // Mode data: 08 01 12 len lang 18 01
    let mut mode_data = vec![0x08, 0x01, 0x12, lang_bytes.len() as u8];
    mode_data.extend_from_slice(lang_bytes);
    mode_data.extend_from_slice(&[0x18, 0x01]);

    // Payload: 08 01 10 msg_id 1A len mode_data
    let mut payload = vec![0x08, 0x01, 0x10, msg_id, 0x1A, mode_data.len() as u8];
    payload.extend(mode_data);

    build_packet(seq, 0x05, 0x20, &payload)
}

/// Build translation disable packet
pub fn build_translation_disable(seq: u8, msg_id: u8) -> Vec<u8> {
    let mode_data = vec![0x08, 0x02];
    let mut payload = vec![0x08, 0x01, 0x10, msg_id, 0x1A, mode_data.len() as u8];
    payload.extend(mode_data);

    build_packet(seq, 0x05, 0x20, &payload)
}

/// Translation result
#[derive(Debug, Clone)]
pub struct TranslationResult {
    pub original: String,
    pub translation: String,
    pub is_final: bool,
}

/// Parse translation result from notification data
pub fn parse_translation_result(data: &[u8]) -> Option<TranslationResult> {
    // Find service 05 20 in data
    let mut service_idx = None;
    for i in 0..data.len().saturating_sub(1) {
        if data[i] == 0x05 && data[i + 1] == 0x20 {
            service_idx = Some(i + 2);
            break;
        }
    }

    let payload_start = service_idx?;
    if payload_start + 4 > data.len() {
        return None;
    }

    let payload = &data[payload_start..data.len().saturating_sub(2)];

    if payload.len() < 4 || payload[0] != 0x08 || payload[1] != 0x02 {
        return None;
    }

    let mut original = String::new();
    let mut translation = String::new();
    let mut is_final = false;
    let mut idx = 2;

    while idx < payload.len() {
        let tag = payload[idx];

        if tag == 0x10 {
            idx += 2; // Skip msg_id
        } else if tag == 0x22 {
            idx += 1;
            if idx >= payload.len() {
                break;
            }
            let content_len = payload[idx] as usize;
            idx += 1;
            if idx + content_len > payload.len() {
                break;
            }
            let content = &payload[idx..idx + content_len];
            idx += content_len;

            // Parse content: 0A len original 12 len translation
            let mut cidx = 0;
            if cidx < content.len() && content[cidx] == 0x0A {
                cidx += 1;
                if cidx < content.len() {
                    let orig_len = content[cidx] as usize;
                    cidx += 1;
                    if cidx + orig_len <= content.len() {
                        original = String::from_utf8_lossy(&content[cidx..cidx + orig_len]).to_string();
                        cidx += orig_len;
                    }
                }

                if cidx < content.len() && content[cidx] == 0x12 {
                    cidx += 1;
                    if cidx < content.len() {
                        let trans_len = content[cidx] as usize;
                        cidx += 1;
                        if cidx + trans_len <= content.len() {
                            translation = String::from_utf8_lossy(&content[cidx..cidx + trans_len]).to_string();
                        }
                    }
                }
            }
        } else if tag == 0x18 {
            idx += 2;
        } else if tag == 0x20 {
            idx += 1;
            if idx < payload.len() {
                is_final = payload[idx] == 0x01;
            }
            idx += 1;
        } else if tag == 0x2A {
            idx += 1;
            if idx < payload.len() {
                let spk_len = payload[idx] as usize;
                idx += 1 + spk_len;
            }
        } else {
            idx += 1;
        }
    }

    if original.is_empty() {
        return None;
    }

    Some(TranslationResult {
        original,
        translation,
        is_final,
    })
}

fn get_languages() -> HashMap<&'static str, &'static str> {
    let mut map = HashMap::new();
    map.insert("EN", "English");
    map.insert("CS", "Czech");
    map.insert("HK", "Cantonese (Hong Kong)");
    map.insert("ZH", "Mandarin Chinese");
    map.insert("JA", "Japanese");
    map.insert("KO", "Korean");
    map.insert("ES", "Spanish");
    map.insert("FR", "French");
    map.insert("DE", "German");
    map.insert("IT", "Italian");
    map.insert("PT", "Portuguese");
    map.insert("RU", "Russian");
    map.insert("AR", "Arabic");
    map
}

async fn find_adapter() -> Result<Adapter, Box<dyn std::error::Error>> {
    let manager = Manager::new().await?;
    let adapters = manager.adapters().await?;
    adapters
        .into_iter()
        .next()
        .ok_or("No Bluetooth adapter found".into())
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let args: Vec<String> = env::args().collect();
    let languages = get_languages();

    if args.contains(&"--list".to_string()) || args.contains(&"-l".to_string()) {
        println!("Available language codes:");
        println!("{}", "-".repeat(40));
        for (code, name) in &languages {
            println!("  {}: {}", code, name);
        }
        println!("\nUsage: cargo run --bin translation -- SOURCE TARGET");
        println!("Example: cargo run --bin translation -- CS EN");
        return Ok(());
    }

    if args.len() < 3 {
        println!("Even G2 Translation");
        println!("{}", "=".repeat(40));
        println!("\nUsage:");
        println!("  cargo run --bin translation -- SOURCE TARGET");
        println!("  cargo run --bin translation -- --list");
        println!("\nExamples:");
        println!("  cargo run --bin translation -- CS EN    # Czech to English");
        println!("  cargo run --bin translation -- HK EN    # Cantonese to English");
        return Ok(());
    }

    let source = args[1].to_uppercase();
    let target = args[2].to_uppercase();

    println!("Even G2 Translation");
    println!("{}", "=".repeat(40));
    println!(
        "\nLanguage: {} → {}",
        languages.get(source.as_str()).unwrap_or(&source.as_str()),
        languages.get(target.as_str()).unwrap_or(&target.as_str())
    );

    println!("\nScanning for G2 glasses...");
    let adapter = find_adapter().await?;
    adapter.start_scan(ScanFilter::default()).await?;
    sleep(Duration::from_secs(10)).await;

    let peripherals = adapter.peripherals().await?;

    let mut device: Option<Peripheral> = None;
    for p in &peripherals {
        if let Some(props) = p.properties().await? {
            if let Some(name) = &props.local_name {
                if name.contains("G2") && name.contains("_L_") {
                    device = Some(p.clone());
                    println!("  Found: {}", name);
                    break;
                }
            }
        }
    }

    if device.is_none() {
        for p in &peripherals {
            if let Some(props) = p.properties().await? {
                if let Some(name) = &props.local_name {
                    if name.contains("G2") {
                        device = Some(p.clone());
                        println!("  Found: {}", name);
                        break;
                    }
                }
            }
        }
    }

    let device = match device {
        Some(d) => d,
        None => {
            println!("ERROR: No G2 glasses found!");
            return Ok(());
        }
    };

    device.connect().await?;
    println!("\nConnected!");

    device.discover_services().await?;

    let chars = device.characteristics();
    let write_char = chars
        .iter()
        .find(|c| c.uuid == char_write())
        .ok_or("Write characteristic not found")?;
    let notify_char = chars
        .iter()
        .find(|c| c.uuid == char_notify())
        .ok_or("Notify characteristic not found")?;

    device.subscribe(notify_char).await?;

    // Authenticate
    println!("\nAuthenticating...");
    for pkt in build_auth_packets() {
        device
            .write(write_char, &pkt, WriteType::WithoutResponse)
            .await?;
        sleep(Duration::from_millis(100)).await;
    }
    sleep(Duration::from_millis(500)).await;
    println!("  Authenticated!");

    let translation_count = Arc::new(Mutex::new(0u32));
    let count_clone = translation_count.clone();

    // Enable translation
    println!("\nEnabling {}>{} translation...", source, target);
    let enable_pkt = build_translation_enable(0x10, 0x50, &source, &target);
    device
        .write(write_char, &enable_pkt, WriteType::WithoutResponse)
        .await?;

    println!("\nTranslation active! Speak to see results.");
    println!("Press Ctrl+C to stop.\n");

    let mut notification_stream = device.notifications().await?;

    // Handle notifications
    let device_clone = device.clone();
    let write_char_clone = write_char.clone();
    
    tokio::spawn(async move {
        tokio::signal::ctrl_c().await.ok();
        println!("\n\nDisabling translation...");
        let disable_pkt = build_translation_disable(0x11, 0x51);
        let _ = device_clone
            .write(&write_char_clone, &disable_pkt, WriteType::WithoutResponse)
            .await;
        sleep(Duration::from_millis(500)).await;
        let count = *count_clone.lock().unwrap();
        println!("Done! Received {} translation(s).", count);
        std::process::exit(0);
    });

    while let Some(data) = notification_stream.next().await {
        if let Some(result) = parse_translation_result(&data.value) {
            let mut count = translation_count.lock().unwrap();
            *count += 1;

            let final_marker = if result.is_final { "[FINAL]" } else { "[...]" };
            println!("\n{} Original: {}", final_marker, result.original);
            println!("      Translation: {}", result.translation);
        }
    }

    device.disconnect().await?;
    Ok(())
}
